plugins/builtin/target/azure-vmss/plugin/plugin.go

```diff
@@ -203,28 +247,125 @@ func (t *TargetPlugin) calculateDirection(vmssDesired, strategyDesired int64) (i
+				}
 			}
 		}
 	}
 }
+
+// Unfortunately, this requires us to iterate over each VM in the VMSS.
+// This can be upwards of 2000 VMs in a single VMSS, which Uniform is limited to 200.
+// Added a timer for tracking how long it takes to process the instance views.
+// Ran with a concurrency limit of 10 --- avoided api rate limits.
+// 100 VMs took about ~1 second to process.
+// Previously when running each VM sequentially, it took about about 20-30 seconds.
+func (t *TargetPlugin) processInstanceViewFlexible(vms *[]armcompute.VirtualMachineScaleSetVM, resourceGroup string, status *sdk.TargetStatus) {
+
+	// Only used during debugging to see how long it takes to process the instance views.
+	start := time.Now()
+
+	// Early exit if there are no VMs in the VMSS.
+	if len(*vms) == 0 {
+		t.logger.Debug("No VMs found in the VMSS, skipping instance view processing.")
+		return
+	}
+
+	// Mainly debugging when initially running the plugin to see how many VMs are in the VMSS.
+	t.logger.Debug("Total VMs found in the Flexible VMSS", "count", len(*vms))
+
+	// Cancelable context which is later used in the goroutines to stop processing.
+	// Triggers when it finds a VM that shows not ready.
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// Using a WaitGroup to wait for all goroutines to finish.
+	// Using a channel to limit the number of concurrent requests to the Azure API.
+	// Mutex is used to protect the status.Ready field from concurrent writes.
+	// sync.Once is used to ensure that we only set the status to not ready once, context will be cancelled after that.
+	var wg sync.WaitGroup
+	requests := make(chan struct{}, 5)
+	mu := sync.Mutex{}
+	once := sync.Once{}
+
+	// Iterate over each VM in the VMSS and get its instance view.
+	for _, vm := range *vms {
+		wg.Add(1)
+
+		requests <- struct{}{}
+
+		// Using a goroutine to fetch the instance view for each VM.
+		// Previously, this was done sequentially.
+		go func(vm armcompute.VirtualMachineScaleSetVM) {
+
+			defer wg.Done()
+			defer func() { <-requests }()
+
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+
+			// Unlike Uniform, this has to use the VirtualMachinesClient to get the instance view.
+			instanceView, err := t.vm.InstanceView(ctx, resourceGroup, *vm.Name, nil)
+			if err != nil {
+				t.logger.Debug("failed to get instance view for VM", "vm_name", *vm.Name, "error", err)
+				return
+			}
+
+			if len(instanceView.Statuses) == 0 || *instanceView.Statuses[0].Code != "ProvisioningState/succeeded" {
+
+				for _, statusCode := range instanceView.Statuses {
+					if *statusCode.Code != "ProvisioningState/succeeded" {
+						t.logger.Debug("VM instance view not ready", "vm_name", *vm.Name, "status_code", *statusCode.Code)
+
+						once.Do(func() {
+							t.logger.Debug("Setting status to not ready", "vm_name", *vm.Name)
+							mu.Lock()
+							status.Ready = false
+							mu.Unlock()
+							cancel()
```

Comment:
Because we're canceling early here, we should probably put a context check in the outer loop as well (before `wg.Add(1)`. Otherwise we end up spawning potentially 1000s of goroutines that we don't need.